optee-utee-build/src/linker.rs

```diff
@@ -0,0 +1,182 @@
+        println!("cargo:rustc-link-arg=-pie");
+        println!("cargo:rustc-link-arg=-Os");
+        match self.linker_type {
+            LinkerType::CC => println!("cargo:rustc-link-arg=-Wl,--sort-section=alignment"),
+            LinkerType::LD => println!("cargo:rustc-link-arg=--sort-section=alignment"),
+        };
+        let mut dyn_list = File::create(out_dir.join("dyn_list"))?;
+        write!(
+            dyn_list,
+            "{{ __elf_phdr_info; trace_ext_prefix; trace_level; ta_head; }};\n"
+        )?;
+        match self.linker_type {
+            LinkerType::CC => println!("cargo:rustc-link-arg=-Wl,--dynamic-list=dyn_list"),
+            LinkerType::LD => println!("cargo:rustc-link-arg=--dynamic-list=dyn_list"),
+        }
+
+        Ok(())
+    }
+}
+
+impl Linker {
+    // generate a link script file for cc/ld, and link to it
+    fn write_and_set_linker_script(
+        &self,
+        out_dir: PathBuf,
+        ta_dev_kit_dir: PathBuf,
+    ) -> Result<(), Error> {
+        const ENV_TARGET_TA: &str = "TARGET_TA";
+        println!("cargo:rerun-if-env-changed={}", ENV_TARGET_TA);
+        let mut aarch64_flag = true;
+        match env::var(ENV_TARGET_TA) {
+            Ok(ref v) if v == "arm-unknown-linux-gnueabihf" || v == "arm-unknown-optee" => {
+                match self.linker_type {
+                    LinkerType::CC => println!("cargo:rustc-link-arg=-Wl,--no-warn-mismatch"),
+                    LinkerType::LD => println!("cargo:rustc-link-arg=--no-warn-mismatch"),
+                };
+                aarch64_flag = false;
+            }
+            _ => {}
+        };
+
+        let f = BufReader::new(File::open(ta_dev_kit_dir.join("src/ta.ld.S"))?);
+        let ta_lds_file_path = out_dir.join("ta.lds");
+        let mut ta_lds = File::create(ta_lds_file_path.clone())?;
+        for line in f.lines() {
+            let l = line?;
+
+            if aarch64_flag {
+                if l.starts_with('#')
+                    || l == "OUTPUT_FORMAT(\"elf32-littlearm\")"
+                    || l == "OUTPUT_ARCH(arm)"
+                {
+                    continue;
+                }
+            } else {
+                if l.starts_with('#')
+                    || l == "OUTPUT_FORMAT(\"elf64-littleaarch64\")"
+                    || l == "OUTPUT_ARCH(aarch64)"
+                {
+                    continue;
+                }
+            }
+
+            if l == "\t. = ALIGN(4096);" {
+                write!(ta_lds, "\t. = ALIGN(65536);\n")?;
+            } else {
+                write!(ta_lds, "{}\n", l)?;
+            }
+        }
+
+        println!("cargo:rustc-link-search={}", out_dir.display());
+        println!("cargo:rerun-if-changed={}", ta_lds_file_path.display());
+        println!("cargo:rustc-link-arg=-T{}", ta_lds_file_path.display());
+        Ok(())
+    }
+
+    fn auto_detect_linker_type() -> LinkerType {
+        const ENV_RUSTC_LINKER: &str = "RUSTC_LINKER";
```

Comment:
Do you expect auto-detection with `RUSTC_LINKER` works when the TA developer has to setup `--config $(LINKER_CFG)` manually? Is there a possibility that we can move that manual linker configuration to this `optee-utee-build` crate too?

Reply:
> Do you expect auto-detection with RUSTC_LINKER works when the TA developer has to setup `--config $(LINKER_CFG)` manually?

Actually the environment `RUSTC_LINKER` indicate what developers setup for linker, it would be empty if TA developer didn't setup `--config $(LINKER_CFG)`, so yes, it would work if TA developer has to setup `--config $(LINKER_CFG)`.

>  Is there a possibility that we can move that manual linker configuration to this optee-utee-build crate too?

No, currently cargo not allow the linker to be changed by build script.